### ISSUE TITLE
Inconsistent best channel assignment

### ISSUE THREAD

**Issue by angelonc** [kwikteam/phy-contrib#145]
We've just started using KiloSort on some custom silicone probes and were noticing some inconsistencies with the "best channel" assigned to each cluster in the phy template-gui. (this issue was also noted in post #110)

I circled one example unit in red here, where the cluster was assigned to 29, but the highest waveform is pretty clearly on channel 24. Additionally, I'm a little confused about how the channels in FeatureView are selected, as it seems to include the PCs from channel 29, but also the ones from 19? Wouldn't it be more useful to show PCs from neighboring electrodes?
![best_channel_issue](https://user-images.githubusercontent.com/3865975/39322020-91401bb6-4956-11e8-958d-2de6eef7fd70.PNG)

Generally, we're curious about how these "best channel" labels are assigned, and whether there is something wrong with our channel map (although we tested this a bit and are fairly confident that this is not the case).

Many thanks!

**Comment by nsteinme** [comment]
The best channel will be chosen as the channel with the peak template
amplitude. Sometimes the shape of the template can mis-match the shape of
the real spikes. Could you check the shape of the template (press W) to
rule out this explanation? The second channel chosen for PCs will just be
the one with the next-largest amplitude, but you can change that selection
by holding ctrl and right-clicking or left-clicking a channel in waveform
view.

**Comment by nsteinme** [comment]
It might be a channel map issue - here's how I think about the channel map: there's some site on the probe that you want to call "channel 0" (perhaps because it is labeled that way on a diagram you have from the manufacturer). The data from that site is in some row of the data file (determined by your acquisition system; let's say it's row 10) and the site has some physical coordinates (let's say x=12 and y=30 in some units). Then 
`chanMap(1)=10; xcoords(1)=12; ycoords(1)=30` 
Where it's zero-indexing in phy but 1-indexing in these variables because you're making them in matlab for kilosort. 

Is that how you did it? If not, that could be the problem. If you already have the channel map correct then this must be some kind of bug (@rossant will have to consult)

**Comment by angelonc** [comment]
Sure, I'm fairly certain that's how I went about it, but I've attached my code below. Basically, I made a flat binary file from Neuralynx recording files, in which the first row is the top channel, second row is the next, and so on (this seems to be confirmed by traceview, as the top channel is our reference, and is flat, and the spikes detected on any channel are both consistent with the spikes in waveform view and seem to be picked up by nearby channels, as would be expected):
![capture](https://user-images.githubusercontent.com/3865975/39398397-2efdf71c-4adb-11e8-8f08-e070078ce279.PNG)

Then created coordinates and mapped the channel ID to them, like so (this code will plot the physical layout, and the mapped channel):
`Nchannels = 32;
connected = true(Nchannels, 1);
chanMap   = [1 3 6 9 12 15 18 21 24 27 30 32 2:3:29 4:3:31];
chanMap0ind = chanMap - 1;

% triode arrangement
connector = 91;
radius = sqrt(121)/2;
triodedist = 25;
angle = 60;
xoffset = sqrt(triodedist^2-(triodedist/2)^2);
yoffset = triodedist/2;

x1 = zeros(1,12);
y1 = [0:-(connector+radius):-11*(connector+radius)];
x2 = ones(1,20)*xoffset;
y2 = y1(2:end-1) + yoffset;
y3 = y1(2:end-1) - yoffset;

x = [x1 x2];
y = [y1 y2 y3];

hold on
for i = 1:length(x)
    plot(x(i),y(i),'r.');
    text(x(i),y(i),num2str(chanMap(i)));
end
hold off`

Basically, I'm confused as to how the labels of the traceview and waveform view are consistent (and it also looks like the depth values are also consistent!), but not the channels used for the clusters. Do they not use the same map?

Thanks again for your help!

**Comment by nsteinme** [comment]
Ok, I don't know. One thing that shouldn't make a difference but try real
quick: could you make all the y-coords positive? (c =
readNPY('channel_positions.npy'); c(:,2)=-c(:2); writeNPY(c,
'channel_positions.npy');)

Otherwise, @rossant?

Btw, common average referencing will help a lot to clean up the data you
show there.

**Comment by angelonc** [comment]
Just tried that, and the issue is the same, but the waveform and traceviews are flipped:
![best_channel_flipped](https://user-images.githubusercontent.com/3865975/39398734-f43066c4-4adf-11e8-9d21-4b0e8c49f685.PNG)

I totally agree about the common average reference, I currently do that on the openEphys acquisition system I use. This is actually some data recorded by other people in the lab on an older system, and I'm trying to setup a pipeline to get it through KiloSort and phy. I should definitely implement a common average reference when I convert it.

**Comment by angelonc** [comment]
It appears that there is an inconsistency in how the cluster/similarity/feature views use the channel map, and how trace/waveform view use it. I attempted to remap my channels, such that their spatial position (top to bottom) was consistent with their mapping order (ie. x,y coordinates for the topmost channel correspond to channel map 1, x,y coordinates for the next channel correspond to channel_map(2))

Mapping it in this fashion makes the cluster labels consistent with the largest channel in waveform view, but now the spatial layout no longer makes sense:
![best_channel_transformed](https://user-images.githubusercontent.com/3865975/39530407-e88d3efe-4df6-11e8-8aa0-f2cea9e720b4.PNG)

I will note that instead of running kilosort again, I just edited the channel_map.npy and channel_positions.npy files. Would this make a difference here?

**Comment by LauriTK** [comment]
I also noticed same kind of inconsistency with the channel numbering between the different views in phy. We are loading openEphy files directly to the Kilosort and our channel order is following:

"""
chanMap = [5 6 7 8 2 3 4 9 1 10 11 32 12 13 30 31 14 15 16 29 17 18 19 28 20 21 26 27 22 23 24 25 ... 
    37 38 39 40 34 35 36 41 33 42 43 64 44 45 62 63 46 47 48 61 49 50 51 60 52 53 58 59 54 55 56 57 ];
"""

In phy, there is no logic with the clusters channel number and with the best channels view in the waveformview window. Remapping in running order might fix this (?).

**Comment by rossant** [comment]
should be fixed in the dev branch of phy
